app/services/weather.py

```
@@ -0,0 +1,329 @@
+            raise RuntimeError("Weather settings are missing in configuration.")
+        w = settings.weather
+        if not w.api_key:
+            logger.warning("OPENWEATHER_API_KEY is not set. Weather will be disabled.")
+
+        self._api_key = w.api_key or ""
+        self._base_url = w.base_url.rstrip("/")
+        self._units = w.units
+        self._timeout = w.request_timeout_seconds
+        self._retry_attempts = w.retry_attempts
+        self._retry_min = w.retry_backoff_min
+        self._retry_max = w.retry_backoff_max
+        global _GLOBAL_CACHE
+        if _GLOBAL_CACHE is None:
+            _GLOBAL_CACHE = _TTLCache(w.cache_ttl_seconds)
+        self._cache = _GLOBAL_CACHE
+
+    async def _get_city_coordinates(self, session: aiohttp.ClientSession, city: str) -> Optional[Tuple[float, float]]:
+        params = {"q": city, "appid": self._api_key, "limit": 1}
+        url = f"{self._base_url}/geo/1.0/direct"
+        async with session.get(url, params=params, timeout=self._timeout) as resp:
+            if resp.status != 200:
+                text = await resp.text()
+                logger.warning(f"Geocoding failed for '{city}': {resp.status} {text}")
+                return None
+            data = await resp.json()
+            if not data:
+                return None
+            entry = data[0]
+            return float(entry.get("lat", 0.0)), float(entry.get("lon", 0.0))
+
+    async def _fetch_forecast_raw(self, session: aiohttp.ClientSession, lat: float, lon: float) -> Dict[str, Any]:
+        params = {
+            "lat": lat,
+            "lon": lon,
+            "appid": self._api_key,
+            "units": self._units,
+        }
+        url = f"{self._base_url}/data/2.5/forecast"
+        async with session.get(url, params=params, timeout=self._timeout) as resp:
+            if resp.status != 200:
+                text = await resp.text()
+                raise RuntimeError(f"OpenWeather forecast error: {resp.status} {text}")
```

Error messages from the weather API are swallowed in generic exceptions, losing information for debugging. Return or log the actual error content.
```
if resp.status != 200:
    text = await resp.text()
    logger.error(f"Error {resp.status} from weather API: {text}")
    raise RuntimeError(f"OpenWeather forecast error: {resp.status} {text}")
```